exporters/zipkin/src/main/java/io/opentelemetry/exporters/zipkin/ZipkinSpanExporter.java

```diff
@@ -0,0 +1,206 @@
+      }
+    }
+    return builder.build();
+  }
+
+  static Span generateSpan(SpanData spanData, Endpoint localEndpoint) {
+    long startTimestamp = toEpochMicros(spanData.getStartEpochNanos());
+
+    long endTimestamp = toEpochMicros(spanData.getEndEpochNanos());
+
+    Span.Builder spanBuilder =
+        Span.newBuilder()
+            .traceId(spanData.getTraceId().toLowerBase16())
+            .id(spanData.getSpanId().toLowerBase16())
+            .kind(toSpanKind(spanData))
+            .name(spanData.getName())
+            .timestamp(toEpochMicros(spanData.getStartEpochNanos()))
+            .duration(endTimestamp - startTimestamp)
+            .localEndpoint(localEndpoint);
+
+    if (spanData.getParentSpanId().isValid()) {
+      spanBuilder.parentId(spanData.getParentSpanId().toLowerBase16());
+    }
+
+    for (Map.Entry<String, AttributeValue> label : spanData.getAttributes().entrySet()) {
+      spanBuilder.putTag(label.getKey(), attributeValueToString(label.getValue()));
+    }
+    Status status = spanData.getStatus();
+    if (status != null) {
+      spanBuilder.putTag(STATUS_CODE, status.getCanonicalCode().toString());
+      if (status.getDescription() != null) {
+        spanBuilder.putTag(STATUS_DESCRIPTION, status.getDescription());
+      }
+      if (!status.isOk()) {
+        spanBuilder.putTag(STATUS_ERROR, status.getCanonicalCode().toString());
+      }
+    }
+
+    for (SpanData.TimedEvent annotation : spanData.getTimedEvents()) {
+      spanBuilder.addAnnotation(toEpochMicros(annotation.getEpochNanos()), annotation.getName());
+    }
+
+    return spanBuilder.build();
+  }
+
+  @Nullable
+  private static Span.Kind toSpanKind(SpanData spanData) {
+    // This is a hack because the Span API did not have SpanKind.
+    if (spanData.getKind() == Kind.SERVER
+        || (spanData.getKind() == null && Boolean.TRUE.equals(spanData.getHasRemoteParent()))) {
+      return Span.Kind.SERVER;
+    }
+
+    // This is a hack because the Span API did not have SpanKind.
+    if (spanData.getKind() == Kind.CLIENT || spanData.getName().startsWith("Sent.")) {
+      return Span.Kind.CLIENT;
+    }
+
+    if (spanData.getKind() == Kind.PRODUCER) {
+      return Span.Kind.PRODUCER;
+    }
+    if (spanData.getKind() == Kind.CONSUMER) {
+      return Span.Kind.CONSUMER;
+    }
+
+    return null;
+  }
+
+  private static long toEpochMicros(long epochNanos) {
+    return MICROSECONDS.convert(epochNanos, NANOSECONDS);
+  }
+
+  private static String attributeValueToString(AttributeValue attributeValue) {
+    Type type = attributeValue.getType();
+    switch (type) {
+      case STRING:
+        return attributeValue.getStringValue();
+      case BOOLEAN:
+        return String.valueOf(attributeValue.getBooleanValue());
+      case LONG:
+        return String.valueOf(attributeValue.getLongValue());
+      case DOUBLE:
+        return String.valueOf(attributeValue.getDoubleValue());
+    }
+    return "";
```

Comment:
usually I add an `assert false` or similar as this is a bug I think. especially in the same repo there should never be an escape from enum switch (to degree it happens outside is avoidable version skew)

Reply:
Probably makes sense to at least add a log message that there was an unknown attribute value type. And, there are array valued types coming...not sure what zipkin wants for array-valued attributes.

Reply:
join on comma if you support array values.

I wouldn't log or if anything log once if there's a bug here on unmatched (avoid tight loops of logging). assert will catch it ideally.

Reply:
This project doesn't use java `assert`, so I don't want to do that. I'd also rather not have the export fail if a new type gets added without the exporter being updated. (like if someone updates their SDK version, but doesn't update their exporter version).

Reply:
up to you. FYI we do it all the time. running assertions in prod isn't usual. but I just wanted to express my opinion and thanks for that. something like a log-once seems fine!

Reply:
ps the reason we do it is if you don't, and only log, the build won't break when something new is added.. and typically logs are ignored. if you think of a better way to handle this case, let me know as that's the only reason we do it (to make sure someone doesn't have to remember all the things that need to be updated)

Reply:
hmmm. if we make sure to have a test that iterates over all the values of the type enum, then we can guarantee the build fails by just having it throw an exception. Since this is a part of the main project, it will fail the build. I'll give this a try tomorrow!

Reply:
I think we have static checks enabled and you fail to compile if we miss a value, no need to have tests for all the values to catch this bug (it is good to have tests for all the values to ensure correctness, but not for this particular problem). Hope I am not wrong, can you give a try?

Reply:
The problem comes when we add a new type, and users will use the API with the new type and an old exporter package. That's when we may have troubles.

Reply:
fair enough!

Reply:
> I think we have static checks enabled and you fail to compile if we miss a value, no need to have tests for all the values to catch this bug (it is good to have tests for all the values to ensure correctness, but not for this particular problem). Hope I am not wrong, can you give a try?

Confirmed. errorprone catches this and fails the build if you're missing an enum in your switch branches.

Reply:
We can simply use throw if you want. No need to use assert

Reply:
changed to throw an exception for now.